### PR TITLE
Improve uploader s3 security

### LINKED ISSUE #2641: Uploader with S3 security concerns
The way uploader has been implemented at [](https://github.com/thecodingmachine/workadventure/blob/d0b4d48706c600f3fb484ee590210f198144c609/uploader/src/Service/UploaderService.ts#L63) presents a serious security concern as it makes the S3 bucket public readable.

Uploader  should not mess up with the bucket ACL and instead use the S3 bucket as is. Uploader already has the AWS credentials, so it can fetch the file from the bucket to serve to the user, the same way you do for redis, instead of returning the URL in the bucket. That way the bucket can stay private.

### PR DESCRIPTION
**Pull request by taksan**
Fixes #2641 

Implements a new approach that generates pre signed S3 URLs for bucket objects on demand, expiring after 60 seconds. This way, the bucket doesn't need to be public, which can be a restriction imposed by compliance in several places.

### CODE REVIEW

**Comment by fcecagno** [comment]
Nicely done, really like the idea of using signed URLs to access objects. I'd suggest to make the expiration time configurable in the future.

**Review by moufmouf — APPROVED** [review]
Woah, this is really excellent.

I love the way you refactored the uploaders in different classes, and the signed-urls feature is just excellent. Plus the documentation is great. I only added a comment on a small typo but apart from that, I'm :+1: on merging.

@gparant Any comments before we merge?

File: uploader/README.md
Comment (taksan) [reply uploader/README.md]: Fixed

**Comment by taksan** [comment]
@moufmouf anything missing for this merge request?

**Comment by moufmouf** [comment]
Hey @taksan!

The PR looks really good. It is missing nothing except some time on our side to properly test it.
We'll do our best to review it quickly. Things are just a bit hectic right now :D

**Comment by moufmouf** [comment]
Hey @taksan,

I just did a few tests, I have 2 comments.

First, I'm testing with Minio, so results might be slightly different than with S3.

I'm getting:

```
ploader_1          | Could not setup CORS for S3 bucket MalformedXML: The XML you provided was not well-formed or did not validate against our published schema.
uploader_1          |     at Request.extractError (/usr/src/app/node_modules/aws-sdk/lib/services/s3.js:711:35)
uploader_1          |     at Request.callListeners (/usr/src/app/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
uploader_1          |     at Request.emit (/usr/src/app/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
uploader_1          |     at Request.emit (/usr/src/app/node_modules/aws-sdk/lib/request.js:686:14)
uploader_1          |     at Request.transition (/usr/src/app/node_modules/aws-sdk/lib/request.js:22:10)
uploader_1          |     at AcceptorStateMachine.runTo (/usr/src/app/node_modules/aws-sdk/lib/state_machine.js:14:12)
uploader_1          |     at /usr/src/app/node_modules/aws-sdk/lib/state_machine.js:26:10
uploader_1          |     at Request.<anonymous> (/usr/src/app/node_modules/aws-sdk/lib/request.js:38:9)
uploader_1          |     at Request.<anonymous> (/usr/src/app/node_modules/aws-sdk/lib/request.js:688:12)
uploader_1          |     at Request.callListeners (/usr/src/app/node_modules/aws-sdk/lib/sequential_executor.js:116:18) {
uploader_1          |   code: 'MalformedXML',
uploader_1          |   region: 'eu-west-1',
uploader_1          |   time: 2022-10-03T17:10:06.436Z,
uploader_1          |   requestId: '171A9E3F2BE80D76',
uploader_1          |   extendedRequestId: undefined,
uploader_1          |   cfId: undefined,
uploader_1          |   statusCode: 400,
uploader_1          |   retryable: false,
uploader_1          |   retryDelay: 60.09870843509124
uploader_1          | }
```

This is because the UPLOADER_AWS_SIGNED_URL_EXPIRATION env var is a string and it needs to be cast to a number.
When I remove the env var, it defaults to 60 (as a number) and everything works. (so gg!)

The second comment is that I think the AWS_URL environment variable is now completely useless, thanks to your PR.
So you can probably remove any reference to it in the code.

**Comment by taksan** [comment]
Fixes pushed

File: uploader/src/Service/S3StorageProvider.ts
Comment (gparant) [inline uploader/src/Service/S3StorageProvider.ts:93]: why not use only `get` implementation and have your contract with only get method?

File: uploader/src/Service/UploaderService.ts
Comment (gparant) [inline uploader/src/Service/UploaderService.ts:14]: Why we need `@ts-ignore`?
If it's mandatory, maybe add comment?

**Comment by taksan** [comment]
@moufmouf Based on your last comment on reusing "get", for that it wouldn't be good to reuse it because of different return in each implementation. So, to reduce the interface and make it more consistent I decided on a bigger refactoring so both storage providers were consistent and had the correct responsibilities. Because it was a big one and I noticed the code I was moving was actually duplicated, I decided to create some e2e tests for the api, including firing up redis and localstack. 

It is now ready for review :)

**Comment by moufmouf** [comment]
Allright!

I do like this!
@gparant Waiting for your green light.

On a totally unrelated note, I started by putting the "uploader" into a separate container because it was a component that was not scaling. At the time, the files were uploaded in memory and directly served from memory. But with the new version by @gparant and @taksan , the uploader "scales", since files are stored in Redis or S3. And if it scales, there is no reason to keep it in another container. I mean, we could take all the code and put it in the pusher. I'm not 100% sure how the pusher would deal with real-time messages + file uploads, but it could simplify the architecture. @gparant  What do you think about this idea? (not to be done now!)

**Comment by taksan** [comment]
@moufmouf @gparant Conflicts resolved. I believe this is ready to go

**Comment by moufmouf** [comment]
Ok, let's do this!
Thanks a lot @taksan ! This is really much appreciated.

**Pull request merged**
